backend/de.metas.business/src/main/java/de/metas/project/budget/BudgetProjectRepository.java

```diff
@@ -78,12 +97,97 @@ public static Optional<BudgetProject> fromRecord(final I_C_Project record)
+												  .projectId(projectId)
+												  .budgets(budgetProjectResourceRepository.getResourcesAsListByProjectId(projectId))
+												  .build())
 						.build());
 	}
+
+	@NonNull
+	public Optional<BudgetProject> getOptionalBy(@NonNull final ProjectQuery query)
+	{
+		final IQueryBuilder<I_C_Project> queryBuilder = queryBL.createQueryBuilder(I_C_Project.class)
+				.addOnlyActiveRecordsFilter()
+				.addInArrayFilter(I_C_Project.COLUMNNAME_AD_Org_ID, query.getOrgId(), OrgId.ANY)
+				.orderByDescending(I_C_Project.COLUMNNAME_AD_Org_ID);
+
+		if (EmptyUtil.isNotBlank(query.getValue()))
+		{
+			queryBuilder.addEqualsFilter(I_C_Project.COLUMNNAME_Value, query.getValue().trim());
+		}
+		if (query.getExternalProjectReference() != null)
+		{
+			queryBuilder.addEqualsFilter(I_C_Project.COLUMNNAME_C_Project_Reference_Ext, query.getExternalProjectReference().getValue());
+		}
+
+		final List<I_C_Project> projectRecords = queryBuilder.create().list();
+		if (projectRecords.isEmpty())
+		{
+			return Optional.empty();
+		}
+
+		return fromRecord(CollectionUtils.singleElement(projectRecords));
+	}
+
+	@NonNull
+	public BudgetProject save(@NonNull final UpsertBudgetProjectRequest upsertBudgetProjectRequest)
+	{
+		final I_C_Project projectRecord = InterfaceWrapperHelper.loadOrNew(upsertBudgetProjectRequest.getProjectId(), I_C_Project.class);
+
+		final BudgetProjectData budgetProjectData = upsertBudgetProjectRequest.getBudgetProjectData();
+
+		projectRecord.setName(budgetProjectData.getName());
+		projectRecord.setValue(budgetProjectData.getValue());
+		projectRecord.setC_ProjectType_ID(ProjectTypeId.toRepoId(budgetProjectData.getProjectTypeId()));
+		projectRecord.setIsActive(budgetProjectData.isActive());
+		projectRecord.setC_Currency_ID(CurrencyId.toRepoId(budgetProjectData.getCurrencyId()));
+		projectRecord.setAD_Org_ID(OrgId.toRepoId(budgetProjectData.getOrgId()));
+		projectRecord.setDescription(budgetProjectData.getDescription());
+		projectRecord.setC_Project_Parent_ID(ProjectId.toRepoId(budgetProjectData.getProjectParentId()));
+		projectRecord.setC_BPartner_ID(BPartnerId.toRepoId(budgetProjectData.getBPartnerId()));
+		projectRecord.setM_PriceList_Version_ID(PriceListVersionId.toRepoId(budgetProjectData.getPriceListVersionId()));
+		projectRecord.setSalesRep_ID(UserId.toRepoId(budgetProjectData.getSalesRepId()));
+		projectRecord.setDateContract(TimeUtil.asTimestamp(budgetProjectData.getDateContract()));
+		projectRecord.setDateFinish(TimeUtil.asTimestamp(budgetProjectData.getDateFinish()));
+		projectRecord.setC_Project_Reference_Ext(budgetProjectData.getProjectReferenceExt());
+
+		saveRecord(projectRecord);
+
+		final ProjectId projectId = ProjectId.ofRepoId(projectRecord.getC_Project_ID());
+
+		upsertBudgetProjectRequest.getProjectResourceRequests()
+				.forEach(upsertBudgetProjectResourceRequest -> budgetProjectResourceRepository.save(upsertBudgetProjectResourceRequest, projectId));
```

Comment:
here we have SQL N+1 performance issue.
(if you are not familiar, a description of that can be found here: https://vladmihalcea.com/n-plus-1-query-problem/ ... and in many other places).

Basically, if your `upsertBudgetProjectRequest.getProjectResourceRequests()` has 100 items, it means you will issue 100 SELECTs to load the database record.

How to solve it?
* send the whole list `upsertBudgetProjectRequest.getProjectResourceRequests()` to budgetProjectResourceRepository.save(..). You can even rename it as `saveAll` or better `createOrUpdateAll` because basically that's what it does
* The new `createOrUpdateAll` shall:
   * load all existing records all together (in one select)
   * iterate the upserts and update existing record if any or create a new record
   * save each record, one by one (on save side, we don't have batching... so saving one by one it's the only option :( ).

Here is an example on how to do that: https://github.com/metasfresh/metasfresh/blob/inner_silence_uat/backend/de.metas.business/src/main/java/de/metas/costrevaluation/CostRevaluationRepository.java#L76-L104

cc @metas-ts